### PR TITLE
finegrained-fp8: add swiglu support

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
## Important: Read before submitting

> **New contributors:** Please open or comment on an issue **before** submitting a
> PR to discuss the change you'd like to make. This helps us align on approach and
> avoids wasted effort on changes we may not be able to merge. Please only create
> a PR once one of the project maintainers agrees on your outlined approach.
>
> PRs of contributors who are not vouched for are automatically closed. Regular
> contributors are added to the vouch list.
>
> The kernels-community repository is for:
>
> 1. Kernels developed by Hugging Face and partners.
> 2. Kernels developed by third parties, that have not been 'kernelized' yet, but
>    are used by Hugging Face projects such as diffusers and transformers.
>
> We cannot accept PRs for the following:
>
> - New kernels that are not in categories (1) and (2).
> - Changes to kernels in category (2), these should be submitted upstream.
>
> For, LLM-generated changes, we prefer that you write your prompt in an issue
> over a PR with LLM-generated changes.

## Related issue

Closes #

## Changes

-

## Testing

-

## Checklist

- [ ] This PR is linked to an issue that was discussed and approved
- [ ] I have tested these changes locally
- [ ] New/changed functionality has test coverage
- LLM disclosure:
  - [ ] I did not use an LLM to create this PR.
  - [ ] I used and LLM for assistance while creating this PR.
  - [ ] This PR was mostly or completely generated by an LLM.

### CODE REVIEW

File: finegrained-fp8/tests/test_moe.py [hunk 1]
```diff
@@ -658,6 +711,19 @@ def _make_moe_inputs(problem: MoEProblem):
...
+    """Reference GLU activation, mirroring the fused kernel's ``glu`` and the model
+    ``_apply_gate``. ``swiglu_limit`` clamps gate above / up to ``[-limit, limit]``;
+    ``swiglu_alpha`` gives clamped SwiGLU-OAI ``(up + 1) * gate * sigmoid(alpha * gate)``
+    (GPT-OSS / MiniMax-M3). Otherwise plain GLU ``act_fn(gate) * up``."""
+    if swiglu_limit is not None:
+        gate = gate.clamp(max=swiglu_limit)
+        up = up.clamp(min=-swiglu_limit, max=swiglu_limit)
+    if swiglu_alpha is not None:
+        return (up + 1.0) * (gate * torch.sigmoid(gate * swiglu_alpha))
+    return _ACT_FNS[act_fn](gate) * up
```
Comment (vasqu) [inline finegrained-fp8/tests/test_moe.py:738]: Ok so we allow clamping separate from the alpha? That's great!

Imo, the only thing we should probably also support is interleaved vs concatenated?

File: finegrained-fp8/tests/test_moe.py [hunk 2]
```diff
@@ -617,6 +626,50 @@ def id(self):
         num_top_k=8,
         block_size=(128, 128),
     ),
+    # ── Clamped SwiGLU-OAI (GPT-OSS / MiniMax-M3) on MXFP4 + MXFP8 ──
```
Comment (vasqu) [inline finegrained-fp8/tests/test_moe.py]: Should we have problem cases for only limit, only alpha?

File: finegrained-fp8/torch-ext/finegrained_fp8/fused_batched.py [hunk 3]
```diff
@@ -655,9 +677,10 @@ def _mxfp_dynamic_moe_batched(
     down_proj_scale: torch.Tensor,
     top_k_index: torch.Tensor,
     top_k_weights: torch.Tensor,
-    output_dtype: torch.dtype,
```
Comment (vasqu) [inline finegrained-fp8/torch-ext/finegrained_fp8/fused_batched.py:658]: just as a nudge, let's be careful to not break signature of public facing API. I havent properly checked but maybe we need to reorder in some cases

File: finegrained-fp8/torch-ext/finegrained_fp8/fused_grouped.py [hunk 4]
```diff
@@ -204,7 +204,7 @@ def _grouped_routing(expert_ids: torch.Tensor, num_experts: int, num_top_k: int)
 # three memory modes; get_mxfp_autotuning_configs drops the device-inapplicable descriptor
 # flavor (host on XPU, device on CUDA).
 _MX_COMPUTE_MODES = ("dot_scaled", "dot")
-_MX_MEMORY_MODES = ("host_descriptor", "device_descriptor", "pointer")
+_MX_MEMORY_MODES = ("descriptor", "pointer")
```
Comment (vasqu) [inline finegrained-fp8/torch-ext/finegrained_fp8/fused_grouped.py:207]: Hmm what happened here?

File: finegrained-fp8/torch-ext/finegrained_fp8/utils.py [hunk 5]
```diff
@@ -278,16 +278,17 @@ def get_mxfp_autotuning_configs(
...
-    the device-inapplicable descriptor flavor (host-side is NVIDIA-TMA-only, device-side is the
-    XPU win). a given pre_hook is attached to every config and self-guards on ``MEMORY_MODE``."""
-    drop = (
-        "host_descriptor" if get_active_device_type() == "xpu" else "device_descriptor"
+    ``memory_modes`` (fused_grouped only) is the ``MEMORY_MODE`` axis for the weight load: pass
+    ``("descriptor", "pointer")`` — ``"descriptor"`` resolves here to the device's tensor-descriptor
+    flavor (host-built / NVIDIA-TMA on CUDA, device-built in-kernel tensormap on XPU), ``"pointer"``
+    is explicit pointers. ``(None,)`` (default) emits no axis. A given pre_hook is attached to every
+    config and self-guards on ``MEMORY_MODE``."""
+    # "descriptor" -> the device's descriptor flavor: host-built (NVIDIA TMA) on CUDA, device-built
```
Comment (vasqu) [inline finegrained-fp8/torch-ext/finegrained_fp8/utils.py:286]: Ah ok this here is the change 👀

**Review by vasqu — COMMENTED** [review]
Just some smaller comments overall. I was thinking mostly about concatenated vs interleaved layout BUT unsure how hard it is to do in triton 🤔 

Edit: I only noticed it because of GPT OAI Priv filter but we only use interleaved tbh so not high prio

File: finegrained-fp8/torch-ext/finegrained_fp8/fused_batched.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/torch-ext/finegrained_fp8/fused_batched.py:658]: these were just shipped in v4, still not in use, allowing output_dtype was a mistake because it's not in all signatures

File: finegrained-fp8/torch-ext/finegrained_fp8/fused_grouped.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/torch-ext/finegrained_fp8/fused_grouped.py:207]: just to simplify, instead of having both and dropping one based on device, we keep descriptor and decide which kind based on device

File: finegrained-fp8/torch-ext/finegrained_fp8/utils.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/torch-ext/finegrained_fp8/utils.py:286]: yes, imo simpler

File: finegrained-fp8/tests/test_moe.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/tests/test_moe.py]: i don't think necessary but can do yes

File: finegrained-fp8/tests/test_moe.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/tests/test_moe.py:738]: would be complicated (need different loads in all kernels), what we can do is apply weight conversion to models with interleaved weights to be able to use them easily. last time with gpt oss + sonicmoe the problem was the limit/alpha (small logits deviation)

File: finegrained-fp8/tests/test_moe.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply finegrained-fp8/tests/test_moe.py]: done

File: finegrained-fp8/tests/test_moe.py (hunk 1 above)
Comment (vasqu) [reply finegrained-fp8/tests/test_moe.py:738]: Yea I also noticed afterwards that this is too much of an exception. It only happens on gpt oss atm re interleaved. Otherwise, we have always concatenated iirc

**Review by vasqu — APPROVED** [review]
Yep thanks!
